jobs/base_model_validator.py

```diff
@@ -0,0 +1,160 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Beam PTransform and DoFn for base model validator functions"""
+
+from __future__ import absolute_import  # pylint: disable=import-only-modules
+from __future__ import unicode_literals  # pylint: disable=import-only-modules
+
+import datetime
+import re
+
+import apache_beam as beam
+
+from core.domain import cron_services
+from core.platform import models
+from jobs import base_model_validator_errors as errors
+from jobs import utils as beam_utils
+
+(base_models, user_models) = models.Registry.import_models(
+    [models.NAMES.base_model, models.NAMES.user])
+
+
+MAX_CLOCK_SKEW_SECS = datetime.timedelta(seconds=1)
+
+
+class ValidateModelIdWithRegex(beam.DoFn):
+    """DoFn to validate model ids against a given regex string."""
+
+    def __init__(self, regex_string):
+        """Initializes the ValidateModelIdWithRegex DoFn.
+
+        Args:
+            regex_string: str. A regex pattern for valid ids.
+        """
+        super(ValidateModelIdWithRegex, self).__init__()
+        self.regex = re.compile(regex_string)
+
+    def process(self, input_model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+            input_model: datastore_services.Model. Entity to validate.
+
+        Yields:
+            ModelInvalidIdError. An error class for models with invalid IDs.
+        """
+        model = beam_utils.clone_model(input_model)
+
+        if not self.regex.match(model.id):
+            yield errors.ModelInvalidIdError(model)
+
+
+class ValidateDeleted(beam.DoFn):
+    """DoFn to check whether models marked for deletion are stale."""
+
+    def process(self, input_model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+            input_model: datastore_services.Model. Entity to validate.
+
+        Yields:
+            ModelExpiredError. An error class for expired models.
+        """
+        model = beam_utils.clone_model(input_model)
+        date_now = datetime.datetime.utcnow()
+
+        expiration_date = (
+            date_now -
+            cron_services.PERIOD_TO_HARD_DELETE_MODELS_MARKED_AS_DELETED)
+
+        if model.last_updated < expiration_date:
+            yield errors.ModelExpiredError(model)
+
+
+class ValidateModelTimeFields(beam.DoFn):
+    """DoFn to check whether created_on and last_updated timestamps are
+    valid."""
+
+    def process(self, input_model):
+        """Function that defines how to process each element in a pipeline of
+        models.
+
+        Args:
+            input_model: datastore_services.Model. Entity to validate.
+
+        Yields:
+            ModelMutatedDuringJobError. Error for timestamp validation.
+            ModelTimestampRelationshipError. Error for timestamp validation.
+        """
+
```

Comment:
nit: please remove this line to stay consistent with other `process` definitions.

Reply:
done